### PR TITLE
Parametrize fit hook test with manual optimization

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Related to the Lightning hooks review: https://github.com/PyTorchLightning/pytorch-lightning/issues/7740

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 1]
```diff
@@ -75,15 +75,7 @@ def backward(
         closure_loss = scaled_loss.__enter__()
 
         # do backward pass
-        # TODO: not entirely sure, why we need this
-        if model is not None and isinstance(model, LightningModule):
-            model.backward(closure_loss, optimizer, opt_idx, **kwargs)
```
Comment (tchaton) [inline pytorch_lightning/plugins/precision/apex_amp.py:80]: This is a BC change right ?

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 2]
```diff
@@ -129,10 +121,6 @@ def pre_optimizer_step(
         """
         # apex amp does not support closures.
         lambda_closure()
-
-        if not pl_module.automatic_optimization:
```
Comment (tchaton) [inline pytorch_lightning/plugins/precision/apex_amp.py:133]: Did we decide to remove the hooks in manual optimization ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/plugins/precision/apex_amp.py:80]: Was just trying things. This PR is a draft - will revert this

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/plugins/precision/apex_amp.py:133]: Nope, will be kept. These changes will be reverted

**Review by Borda — APPROVED** [review]
lgtm

**Comment by carmocca** [comment]
> lgtm, really 74 commits? can you prune the merge message later...

Yeah, I always remove the commit list when merging

**Review by kaushikb11 — APPROVED** [review]

File: tests/models/test_hooks.py [hunk 3]
```diff
@@ -322,6 +329,36 @@ def _train_batch(trainer, model, batches, device=torch.device('cpu'), current_ep
...
+            out.extend([
+                dict(name='on_before_batch_transfer', args=(ANY, 0)),
+                dict(name='transfer_batch_to_device', args=(ANY, device, 0)),
+                dict(name='on_after_batch_transfer', args=(ANY, 0)),
+                # TODO: `on_batch_{start,end}`
+                dict(name='Callback.on_batch_start', args=(trainer, model)),
+                dict(name='Callback.on_train_batch_start', args=(trainer, model, ANY, i, 0)),
+                dict(name='on_train_batch_start', args=(ANY, i, 0)),
+                dict(name='forward', args=(ANY, )),
+                dict(name='optimizers'),
```
Comment (awaelchli) [inline tests/models/test_hooks.py:346]: this is not a hook, and it could be called an arbitrary number of times.

**Review by awaelchli — APPROVED** [review]

File: tests/models/test_hooks.py (hunk 3 above)
Comment (carmocca) [reply tests/models/test_hooks.py]: Removed

**Review by tchaton — APPROVED** [review]
LGTM !
